### ISSUE TITLE
Devtools run like shit.

### ISSUE THREAD

**Issue by wolfbeast** [MoonchildProductions/UXP#338]
The UXP devtools run like absolute shit. They are slow, poorly-responsive, take forever to initialize and are generally unpleasant to work with as a result.

We should see if there's a way to improve their performance.

**Comment by JustOff** [comment]
Yeah, the initialization is rather slow, but on the other hand, now they are not loaded until the first call, that saves the memory (the difference is quite noticeable), that positively affects the browser startup time.

**Comment by mattatobin** [comment]
The best compromise that would allow us to preserve it's lazyloaded on-demand status would be to rewrite the bulk of the fe as proper xul overlays. When the user starts up devtools that will set an observer that will use a bootstrap-style method to dynamically create and destroy the chrome registry for devtools on the fly instead of its UI being completely created by js dom manipulation.

This observer would also fire off a timer when user closes all of devtools and eventually deallocate and destroy the chrome registry for the overlays.

Not saying it won't still be a touch on the laggy side but it will be exceedingly improved.

Another thing that needs to happen is we need to evaluate exactly WHAT components of devtools we actually care about and get rid of the rest.

**Comment by janekptacijarabaci** [comment]
> Another thing that needs to happen is we need to evaluate exactly WHAT components of devtools we actually care about and get rid of the rest.

The only thing that comes to my mind is `WebIDE`. I cannot see other possible part at the moment.

**Comment by mattatobin** [comment]
I think remote debugging needs to go as well.. Only consumers of devtools would ever theoretically be able to use it and I have never personally gotten it to work because at the point where you would actually need it the target application to be debugged is so busted it can't even START the debugging server.

**Comment by wolfbeast** [comment]
`Browser Toolbox` being.. what exactly?

**Comment by mattatobin** [comment]
@janekptacijarabaci You have to forgive us because we develop fe in the old tried and true ways of Toolkit Error Console and DOMi.

Do Explain these wonderful Mozilla innovations that theoretically make our lives better.

**Comment by janekptacijarabaci** [comment]
> "Browser Toolbox" being.. what exactly?

Tools / Debugging (xul, js, css - `Inspector`, `Debugger`, `Style Editor`...) of the browser, for `chrome:`, addons...

I use it practically daily.

**Comment by wolfbeast** [comment]
What about GCLI? it's never gone beyond experimental -- Mozilla is actually planning to remove it too, having been incomplete for years.

**Comment by JustOff** [comment]
Just for the record, I also can't even imagine any development without DevTools.

**Comment by janekptacijarabaci** [comment]
> Mozilla is actually planning to remove it too

Yes, but she just wants to use something similar. And I don't see the reason to remove it when it works (I do not see any problem at the moment).
But still, this decision is not for me, but:

> I also can't even imagine any development without DevTools

I simply would not do anything...

**Comment by mattatobin** [comment]
Well you were all waiting for it.. So let me get it out of the way.. DevTools should be an extension.

Now that that's out of the way, we can get back to what should and can be done to make DevTools not run like shit.
